reps/2024-06-16-support-apache-yunikorn-scheduler.md

```diff
@@ -0,0 +1,157 @@
+each scheduler implementation to inject custom resources and modify pod metadata accordingly. When a Ray cluster is
+created, the framework calls the appropriate scheduler plugin functions based on the scheduler name provided
+by `ray.io/scheduler-name`.
+
+### Limitations
+
+The framework is designed to be scheduler-agnostic and provides general hooks for supporting different scheduler options.
+However, once the `enable-batch-scheduler` option is set to `true`, the scheduler manager will attempt to load all
+scheduler schemas by calling the `AddToScheme(scheme *runtime.Scheme)` function implemented by each scheduler plugin.
+This loads all the schemas, including CRDs defined in the implementation. Since only the `VolcanoBatchScheduler`
+is currently implemented, it always loads Volcano's `PodGroup` CRD in the controller runtime,
+requiring the installation of Volcano CRDs even when enabling other scheduler options.
+
+## Proposed changes
+
+### Targeted scheduler option
+
+To support Apache YuniKorn, and more importantly, to other schedulers in Ray.
+We propose to add an option to explicitly set the scheduler name, i.e `--batch-scheduler-name`.
+Option syntax: `--batch-scheduler-name [SupportedSchedulerName]`, for example:
+
+Option Syntax:
+
+```shell
+# 1. use volcano
+--enable-batch-scheduler=true
+--batch-scheduler-name=volcano
+
+# 2. use yunikorn
+--enable-batch-scheduler=true
+--batch-scheduler-name=yunikorn
+
+# 3. use default Kubernetes scheduler, or do not turn on the feature flag
+--enable-batch-scheduler=true
+--batch-scheduler-name=default
+```
+
+When a scheduler name is specified, the scheduler manager will only load the configured scheduler plugin,
+ensuring only necessary resources are loaded in the controller runtime.
+
+The option `--batch-scheduler-name` accepts a single scheduler name as the value, and the value must be `default`,
+`volcano` and `yunikorn` (before other scheduler plugins introduced). If an unrecognized scheduler name is provided,
+the controller will fail to start with an error indicating that the scheduler plugin is not found.
+
+### Deprecate "ray.io/scheduler-name"
+
+With the scheduler name is specified in the operator startup options, there is no need to set `ray.io/scheduler-name`
+in `RayJob` or `RayCluster` CRs. This option should be marked as deprecated and eventually removed.
+
+### Compatibility
+
+When this change is introduced, existing users who use `--enable-batch-scheduler` with the Volcano scheduler
+must explicitly add `--batch-scheduler-name=volcano`. Otherwise, the controller will default to the Kubernetes scheduler.
+This is an incompatible change and must be carefully documented to mitigate potential impact.
+
+### YuniKorn scheduler plugin behavior
+
+The YuniKorn scheduler plugin will support both `RayJob` and `RayCluster` resources. The integration will be lightweight,
```

Comment:
Supporting RayJob is going to require some larger  change in the batch scheduler framework right? Right now it only supports RayCluster.

Reply:
At least for yunikorn, it doesn't require many changes.
Please see the doc for details. Today if we submit a RayJob, the controller already poplates the RayJob's labels to RayCluster, so the generated RayCluster already has most of the info. The only thing is we probably want to reuse the JobID, no matter if that is provided by the user, or generated by the controller.